File: server/msgd/core/payloads/file.py
"""``file.*`` payload schemas (TDD §2.2 / M3.5 Phase-A).

Same modeling discipline as
:class:`~msgd.core.payloads.message.MessageCreatedV1`:

* ``model_config = ConfigDict(extra="allow")`` so additive-only v1 changes
  (§2.3.2) round-trip losslessly through an older reader.
* **Format-validation only** — ``file_id`` prefix + ULID validity, the content
  hash format, and the bounded ``name`` / ``mime_type`` / ``size_bytes`` domains
  (below).  Referential *existence* (does the blob exist? was it actually
  uploaded?) is a server concern (ENG-116, §3.2), never enforced here.

**LOCKED DECISIONS (§2.2-style — changing any of these ⇒ ``type_version`` bump):**

* ``sha256`` is the blob's content hash as a **bare 64-char lowercase hex**
  string (``^[0-9a-f]{64}$``) — NOT the ``sha256:<hex>`` prefixed form used by
  ``event_hash``.  This deliberately matches the content-addressed BlobStore
  (ENG-115): blobs are keyed by bare-hex sha256, §6 has clients compute
  ``sha256(file)`` as bare hex, and the Files API (ENG-116) matches the streamed
  blob to this field, all bare-hex.  The algorithm is fixed by the field name, so
  no ``sha256:`` prefix is carried.
* ``name`` is a bounded, **opaque** filename: a non-empty Unicode string of at
  most :data:`MAX_FILE_NAME_BYTES` (255) UTF-8 bytes.  It is treated as opaque
  display text — no path/extension/charset interpretation happens here.
* ``mime_type`` is a bounded ``type/subtype`` string: non-empty, at most
  :data:`MAX_MIME_TYPE_BYTES` (255) UTF-8 bytes, exactly one ``"/"`` with a
  non-empty type and subtype.  Deliberately **not** the full RFC 6838 token
  grammar — the payload only gates the coarse shape and the length; richer
  sniffing/allow-listing is a server (ENG-116) concern.
* ``size_bytes`` is a non-negative integer within the JCS interop cap
  (``0 <= size_bytes <= 2**53 - 1``; see :mod:`msgd.core.jcs`).  The **50 MB
  business cap is a SERVER-validation concern (ENG-116), NOT the payload** — the
  frozen payload only bounds it as a non-negative in-range integer so the number
  round-trips byte-identically cross-language.
"""

from __future__ import annotations

import re
from typing import Final

from pydantic import BaseModel, ConfigDict, field_validator

from msgd.core import ids

__all__ = [
    "MAX_FILE_NAME_BYTES",
    "MAX_MIME_TYPE_BYTES",
    "MAX_FILE_SIZE_BYTES",
    "FileUploadedV1",
]

#: Upper bound on the UTF-8 byte length of a ``file.uploaded`` ``name`` (locked
#: at ``type_version`` 1).  255 bytes matches the common filesystem ``NAME_MAX``.
MAX_FILE_NAME_BYTES: Final = 255

#: Upper bound on the UTF-8 byte length of a ``mime_type`` (locked at v1).
MAX_MIME_TYPE_BYTES: Final = 255

#: Upper bound on ``size_bytes`` — the JCS integer interop cap ``2**53 - 1``
#: (:mod:`msgd.core.jcs`).  This is the *protocol* bound; the 50 MB business cap
#: is enforced server-side (ENG-116), not here.
MAX_FILE_SIZE_BYTES: Final = 2**53 - 1

#: Bare 64-char lowercase hex — the content-addressed BlobStore key form (ENG-115),
#: NOT the ``sha256:<hex>`` prefixed form used by ``event_hash``.
_SHA256_RE: Final = re.compile(r"^[0-9a-f]{64}$")

#: Coarse ``type/subtype`` shape: exactly one ``/`` with non-empty, slash-free halves.
_MIME_TYPE_RE: Final = re.compile(r"^[^/]+/[^/]+$")


def _require_file_id(value: str) -> str:
    if not ids.is_valid_typed_id(value, ids.IdKind.FILE):
        raise ValueError(f"file_id is not a valid f_ id: {value!r}")
    return value


def _require_sha256(value: str) -> str:
    if not _SHA256_RE.fullmatch(value):
        raise ValueError(f"sha256 must be 64 lowercase hex chars (bare, no prefix), got {value!r}")
    return value


def _require_name(value: str) -> str:
    if value == "":
        raise ValueError("name must be non-empty")
    n = len(value.encode("utf-8"))
    if n > MAX_FILE_NAME_BYTES:
        raise ValueError(f"name is {n} bytes UTF-8, exceeds the {MAX_FILE_NAME_BYTES}-byte limit")
    return value


def _require_mime_type(value: str) -> str:
    n = len(value.encode("utf-8"))
    if n == 0:
        raise ValueError("mime_type must be non-empty")
    if n > MAX_MIME_TYPE_BYTES:
        raise ValueError(
            f"mime_type is {n} bytes UTF-8, exceeds the {MAX_MIME_TYPE_BYTES}-byte limit"
        )
    if not _MIME_TYPE_RE.fullmatch(value):
        raise ValueError(f"mime_type is not a type/subtype string: {value!r}")
    return value


def _require_size_bytes(value: int) -> int:
    if value < 0:
        raise ValueError(f"size_bytes must be >= 0, got {value}")
    if value > MAX_FILE_SIZE_BYTES:
        raise ValueError(
            f"size_bytes {value} exceeds the interop cap {MAX_FILE_SIZE_BYTES} (2**53 - 1)"
        )
    return value


class FileUploadedV1(BaseModel):
    """Payload for ``file.uploaded`` v1 (§2.2 / M3.5 Phase-A).

    The frozen cross-language descriptor of an uploaded blob: its ``file_id``,
    content ``sha256``, opaque ``name``, ``mime_type``, and ``size_bytes``.  This
    model validates the event *shape* only — blob existence, the 50 MB business
    cap, and dedup-by-content-hash are server concerns (ENG-116), never enforced
    here (see the module docstring).
    """

    model_config = ConfigDict(extra="allow")

    file_id: str
    sha256: str
    name: str
    mime_type: str
    size_bytes: int

    @field_validator("file_id")
    @classmethod
    def _check_file_id(cls, value: str) -> str:
        return _require_file_id(value)

    @field_validator("sha256")
    @classmethod
    def _check_sha256(cls, value: str) -> str:
        return _require_sha256(value)

    @field_validator("name")
    @classmethod
    def _check_name(cls, value: str) -> str:
        return _require_name(value)

    @field_validator("mime_type")
    @classmethod
    def _check_mime_type(cls, value: str) -> str:
        return _require_mime_type(value)

    @field_validator("size_bytes")
    @classmethod
    def _check_size_bytes(cls, value: int) -> int:
        return _require_size_bytes(value)

File: web/src/core/payloads/file.ts
/**
 * `file.*` payload schemas — the browser port of
 * `server/msgd/core/payloads/file.py` (TDD §2.2 / M3.5 Phase-A).
 *
 * Id fields are *format-validated only*: prefix + ULID validity are checked to
 * catch malformed references early. Referential existence (does the blob exist?
 * was it actually uploaded?) is a server-side concern (ENG-116, §3.2), out of
 * scope here.
 *
 * LOCKED DECISIONS (§2.2-style — changing any ⇒ `type_version` bump), mirrored
 * exactly from the Python validators so both languages accept/reject the same
 * values:
 *
 * - `sha256` is the blob's content hash as a bare 64-char lowercase hex string
 *   (`^[0-9a-f]{64}$`) — NOT the `sha256:<hex>` prefixed form used by
 *   `event_hash`. Matches the content-addressed BlobStore key (ENG-115).
 * - `name` is a bounded, opaque filename: non-empty, at most
 *   {@link MAX_FILE_NAME_BYTES} (255) UTF-8 bytes.
 * - `mime_type` is a bounded `type/subtype` string: non-empty, at most
 *   {@link MAX_MIME_TYPE_BYTES} (255) UTF-8 bytes, exactly one `/` with
 *   non-empty halves. Deliberately not the full RFC 6838 token grammar.
 * - `size_bytes` is a non-negative integer within the JCS interop cap
 *   (`0 <= size_bytes <= 2**53 - 1`). The 50 MB business cap is a server
 *   concern (ENG-116), NOT the payload.
 */

import { IdKind, isValidTypedId } from '../ids'

/** Upper bound on the UTF-8 byte length of a `file.uploaded` `name` (locked at v1). */
export const MAX_FILE_NAME_BYTES = 255

/** Upper bound on the UTF-8 byte length of a `mime_type` (locked at v1). */
export const MAX_MIME_TYPE_BYTES = 255

/**
 * Upper bound on `size_bytes` — the JCS integer interop cap `2**53 - 1`
 * (`Number.MAX_SAFE_INTEGER`). The 50 MB business cap is server-side (ENG-116).
 */
export const MAX_FILE_SIZE_BYTES = Number.MAX_SAFE_INTEGER

const utf8 = new TextEncoder()

/** Bare 64-char lowercase hex — the content-addressed BlobStore key form (ENG-115). */
const SHA256_RE = /^[0-9a-f]{64}$/
/** Coarse `type/subtype` shape: exactly one `/` with non-empty, slash-free halves. */
const MIME_TYPE_RE = /^[^/]+\/[^/]+$/

function requireFileId(fileId: string): string {
  if (!isValidTypedId(fileId, IdKind.FILE)) {
    throw new Error(`file_id is not a valid f_ id: ${fileId}`)
  }
  return fileId
}

function requireSha256(sha256: string): string {
  if (!SHA256_RE.test(sha256)) {
    throw new Error(`sha256 must be 64 lowercase hex chars (bare, no prefix), got ${sha256}`)
  }
  return sha256
}

function requireName(name: string): string {
  if (name === '') {
    throw new Error('name must be non-empty')
  }
  const n = utf8.encode(name).length
  if (n > MAX_FILE_NAME_BYTES) {
    throw new Error(`name is ${n} bytes UTF-8, exceeds the ${MAX_FILE_NAME_BYTES}-byte limit`)
  }
  return name
}

function requireMimeType(mimeType: string): string {
  const n = utf8.encode(mimeType).length
  if (n === 0) {
    throw new Error('mime_type must be non-empty')
  }
  if (n > MAX_MIME_TYPE_BYTES) {
    throw new Error(`mime_type is ${n} bytes UTF-8, exceeds the ${MAX_MIME_TYPE_BYTES}-byte limit`)
  }
  if (!MIME_TYPE_RE.test(mimeType)) {
    throw new Error(`mime_type is not a type/subtype string: ${mimeType}`)
  }
  return mimeType
}

function requireSizeBytes(sizeBytes: number): number {
  if (!Number.isInteger(sizeBytes)) {
    throw new Error(`size_bytes must be an integer, got ${sizeBytes}`)
  }
  if (sizeBytes < 0) {
    throw new Error(`size_bytes must be >= 0, got ${sizeBytes}`)
  }
  if (sizeBytes > MAX_FILE_SIZE_BYTES) {
    throw new Error(`size_bytes ${sizeBytes} exceeds the interop cap ${MAX_FILE_SIZE_BYTES}`)
  }
  return sizeBytes
}

/** Payload for `file.uploaded` v1 (§2.2 / M3.5 Phase-A). */
export type FileUploadedV1 = {
  file_id: string
  sha256: string
  name: string
  mime_type: string
  size_bytes: number
}

/** Options for {@link buildFileUploadedPayload}; mirrors the Python model fields. */
export interface BuildFileUploadedPayloadOptions {
  file_id: string
  sha256: string
  name: string
  mime_type: string
  size_bytes: number
}

/**
 * Format-validate a `file.uploaded` v1 payload.
 *
 * Mirrors `FileUploadedV1`'s field validators: `file_id` is an `f_` id,
 * `sha256` is bare 64-char lowercase hex, `name`/`mime_type` are bounded, and
 * `size_bytes` is a non-negative in-cap integer.
 *
 * @throws {Error} on a malformed `file_id`, `sha256`, `name`, `mime_type`, or
 *   an out-of-range `size_bytes`.
 */
export function buildFileUploadedPayload(options: BuildFileUploadedPayloadOptions): FileUploadedV1 {
  return {
    file_id: requireFileId(options.file_id),
    sha256: requireSha256(options.sha256),
    name: requireName(options.name),
    mime_type: requireMimeType(options.mime_type),
    size_bytes: requireSizeBytes(options.size_bytes),
  }
}
